vector/src/main/java/im/vector/app/features/location/LocationSharingServiceConnection.kt

```diff
@@ -49,30 +51,46 @@ class LocationSharingServiceConnection @Inject constructor(
         }
     }
 
-    fun unbind() {
-        callback = null
-    }
-
-    fun stopLiveLocationSharing(roomId: String) {
-        locationSharingService?.stopSharingLocation(roomId)
+    fun unbind(callback: Callback) {
+        removeCallback(callback)
     }
 
     override fun onServiceConnected(className: ComponentName, binder: IBinder) {
         locationSharingService = (binder as LocationSharingService.LocalBinder).getService().also {
             it.callback = this
         }
         isBound = true
-        callback?.onLocationServiceRunning()
+        onCallbackActionNoArg(Callback::onLocationServiceRunning)
     }
 
     override fun onServiceDisconnected(className: ComponentName) {
         isBound = false
         locationSharingService?.callback = null
         locationSharingService = null
-        callback?.onLocationServiceStopped()
+        onCallbackActionNoArg(Callback::onLocationServiceStopped)
     }
 
     override fun onServiceError(error: Throwable) {
-        callback?.onLocationServiceError(error)
+        forwardErrorToCallbacks(error)
+    }
+
+    @Synchronized
+    private fun addCallback(callback: Callback) {
+        callbacks.add(callback)
+    }
+
+    @Synchronized
+    private fun removeCallback(callback: Callback) {
+        callbacks.remove(callback)
+    }
+
+    @Synchronized
+    private fun onCallbackActionNoArg(action: Callback.() -> Unit) {
+        callbacks.forEach(action)
+    }
+
+    @Synchronized
+    private fun forwardErrorToCallbacks(error: Throwable) {
+        callbacks.forEach { it.onLocationServiceError(error) }
```

Comment:
if the entry points and service callbacks are all on the same thread, this doesn't need to be synchronised, (could argue for introducing a main thread handler to enforce a single thread if multiple threads are used!) 

also worth highlighting that `synchronised` doesn't guard against concurrent modifications on the same thread eg a `onLocationServiceError` that calls `unbind`

Reply:
:raised_hands:  

this is a tricky one.... I would lean towards not guarding against it and trying to enforce that adding/removing should always happen outside of the callbacks, but it's only enforceable at the code review point :thinking: 

we could use a `CopyOnWriteArray` or copy the list each time we iterate `callbacks.toList()`, both have their own set of drawbacks~

what would be your preference?

Reply:
Ah I see, thanks! I will go for `toList()`, we won't have large lists so I think it is okay.